src/script/interpreter.cpp

```diff
@@ -1953,6 +2011,58 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
                 break;
 
+                case OP_ECMULSCALAREXPVERIFY:
+                {
+                    // OP_ECMULSCALAREXPVERIFY is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+                    valtype& vchScalar = stacktop(-1);
+                    valtype& vchGenerator = stacktop(-2);
+                    valtype& vchRes = stacktop(-3);
+
+                    CPubKey pk(vchGenerator);
+                    CPubKey res(vchRes);
+                    if (!pk.IsCompressed() || !res.IsCompressed())
+                        return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
+                    execdata.m_validation_weight_left -= VALIDATION_WEIGHT_PER_SIGOP_PASSED;
+                    if (execdata.m_validation_weight_left < 0) {
+                        return set_error(serror, SCRIPT_ERR_TAPSCRIPT_VALIDATION_WEIGHT);
+                    }
+                    if (vchScalar.size() != 32 || !pk.TweakMulVerify(res, uint256(vchScalar)))
+                        return set_error(serror, SCRIPT_ERR_ECMULTVERIFYFAIL);
+
+                    popstack(stack);
+                    popstack(stack);
+                    popstack(stack);
+                }
+                break;
+                case OP_TAPTWEAKVERIFY:
+                {
+                    // OP_TAPTWEAKVERIFY is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    valtype& vchInternalKey = stacktop(-1);
+                    valtype& vchTweak = stacktop(-2);
+                    valtype& vchTweakedKey = stacktop(-3);
+
+                    if (vchTweakedKey.size() != CPubKey::COMPRESSED_SIZE || vchInternalKey.size() != 32 || vchTweak.size() != 32)
+                        return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
+
+                    execdata.m_validation_weight_left -= VALIDATION_WEIGHT_PER_SIGOP_PASSED;
+                    if (execdata.m_validation_weight_left < 0) {
+                        return set_error(serror, SCRIPT_ERR_TAPSCRIPT_VALIDATION_WEIGHT);
+                    }
+                    const XOnlyPubKey tweakedKey{uint256(std::vector<unsigned char>(vchTweakedKey.begin() + 1, vchTweakedKey.begin() + CPubKey::COMPRESSED_SIZE))};
+                    const uint256 tweak(vchTweak);
+                    const XOnlyPubKey internalKey{uint256(vchInternalKey)};
+                    if (!tweakedKey.CheckPayToContract(internalKey, tweak, vchTweakedKey[0] & 1))
```

Comment:
AFAICT there is not check that vchTweakedKey[0] is either 0x02 or 0x03.

More testing?

Reply:
Great catch! This is indeed a malleability vector.

Reply:
This is checked above in the `IsFullyValid` check on `tweakedKey`

Reply:
That was added after Russell pointed this out.